File: compiler/examples/smoke.ex
defmodule Smoke do
  # compile to plain (non-typed) registers so our from-scratch decoder handles it
  @compile [:no_type_opt, :no_line_info]

  def add(a, b), do: a + b
  def dbl(x), do: x * 2

  def fact(0), do: 1
  def fact(n) when n > 0, do: n * fact(n - 1)

  def fib(0), do: 0
  def fib(1), do: 1
  def fib(n) when n > 1, do: fib(n - 1) + fib(n - 2)
end

File: attic/spikes/04-beam-loader-smoketest/smoke.erl
-module(smoke).
-export([add/2, dbl/1, fact/1, fib/1]).

add(A, B) -> A + B.

dbl(X) -> X * 2.

fact(0) -> 1;
fact(N) when N > 0 -> N * fact(N - 1).

fib(0) -> 0;
fib(1) -> 1;
fib(N) when N > 1 -> fib(N - 1) + fib(N - 2).
